Shadow container (#638)
unoplatform/uno.toolkit.ui: doc/controls/ShadowContainer.md

---
uid: Toolkit.Controls.ShadowContainer
---

## Summary
The `ShadowContainer` provides the possibility to add many-colored shadows to its content.

## Remarks

> For now, the control simply adapts its corner radius to the content's corner radius. More complicated shapes like texts or pictures with alpha, are not supported.

Reviewer: ```suggestion
For now, the control simply adapts its corner radius to the content's corner radius. More complicated shapes, such as text or pictures with alpha, are not supported.
```

### XAML
```xml
xmlns:utu="using:Uno.Toolkit.UI"
...

<utu:ShadowContainer>
    <utu:ShadowContainer.Shadows>
        <utu:ShadowCollection>
            <utu:Shadow ... />
            <utu:Shadow ... />
        </utu:ShadowCollection>
    </utu:ShadowContainer.Shadows>

    <SomeControl />

</utu:ShadowContainer>
```

### Inheritance
Object &#8594; DependencyObject &#8594; UIElement &#8594; FrameworkElement &#8594; Control &#8594; ContentControl > ShadowContainer

### Properties
| Property | Type             | Description |
| -------- | ---------------- | ----------- |
Shadows  | ShadowCollection |  The collection of shadows that will be displayed under your control. A `ShadowCollection` can be stored in a resource dictionary to have a consistent style through your app. The `ShadowCollection` implements `INotifyCollectionChanged`.

## Shadow

Dependency object representing a single shadow.

### Shadow Properties
| Property | Type | Description |
| -------- | ---- | ----------- |

> IsInner | bool | True is this shadow is an inner shadow (like `inset` in css box-shadow).

Reviewer: ```suggestion
IsInner | bool | True if this shadow is an inner shadow (similar to `inset` of `box-shadow` in CSS).
```

					HorizontalAlignment="Center"
					Orientation="Horizontal"
					Spacing="16">

			<!-- Reference to the dictionary button shadows -->
			<ui:ShadowContainer Shadows="{StaticResource ButtonShadows}">
				<Button Background="{StaticResource UnoColor}"
						BorderThickness="1"
						Content="Add Shadow"
						Foreground="White" />
			</ui:ShadowContainer>

			<!-- Reference to the dictionary button shadows -->
			<ui:ShadowContainer Shadows="{StaticResource ButtonShadows}">
				<Button Background="{StaticResource UnoColor}"
						BorderThickness="1"
						Content="Remove Shadow"
						Foreground="White" />
			</ui:ShadowContainer>
		</StackPanel>
	</StackPanel>
</ui:ShadowContainer>
```

![2 colored shadows with 2 buttons with single shadows](../assets/shadows-colors.png)

### Neumorphism

[Following neumorphism rules](https://neumorphism.io), choose one background color, 2 shadow colors, and create a cool neumorphism style.
In order to achieve neumorphic depth effects (instead of having a raised element, it will feel as if it was hollow or bulging), set the `IsInner` property of a shadow to `true`. The shadow will then be displayed *inside* the element instead of beyond.

> In css `box-shadow` it's equivalent to the `inset` property.

Reviewer: ```suggestion
This is equivalent to the `inset` property of the CSS `box-shadow`.
```

---
uid: Toolkit.Controls.ShadowContainer
---

> ## Summary
> The `ShadowContainer` provides the possibility to add many colored shadows to its content.
> 
> ## Remarks
> For now the control simply adapts its corner radius to the content's corner radius. More complicated shapes like texts or pictures with alpha, are not supported.

Reviewer 2: ```suggestion
## Summary
The `ShadowContainer` provides the possibility to add many-colored shadows to its content.

## Remarks
For now, the control simply adapts its corner radius to the content's corner radius. More complicated shapes like texts or pictures with alpha, are not supported.
```

Shadows  | ShadowCollection |  The collection of shadows that will be displayed under your control. A `ShadowCollection` can be stored in a resource dictionary to have a consistent style through your app. The `ShadowCollection` implements `INotifyCollectionChanged`.

## Shadow

Dependency object representing a single shadow.
Public properties are all dependency properties.

### Shadow Properties
| Property | Type | Description |
| -------- | ---- | ----------- |
IsInner | bool | True is this shadow is a inner shadow (like `inset` in css box-shadow).
OffsetX | double | The X offset of the shadow.
OffsetY | double | The Y offset of the shadow.
Color | Color | The color of the shadow. It will be multiplied by the `Opacity` property before rendering.
Opacity | double | The opacity of the shadow.
BlurRadius | double | The radius of the blur that will be applied to the shadow **[0..100]**.
Spread | double | The spread will inflate or deflate (if negative) the control shadow size **before** applying the blur.

## Usage

```xml
xmlns:utu="using:Uno.Toolkit.UI"
...

<Page.Resources>
    <Color x:Key="UnoColor">#7a67f8</Color>
    <Color x:Key="UnoPink">#f85977</Color>

>     <!-- You can define your shadows in resource dictionary -->

Reviewer 2: ```suggestion
    <!-- You can define your shadows in the resource dictionary -->
```
